protocol/binary.go

```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	e, err := reader.ReadEnveloped()
 	return e, err
 }
+
+// DecodeRequest specializes Decode and replaces DecodeEnveloped for the
+// specific purpose of decoding request structs that may or may not have an
+// envelope.
+// This allows a Thrift request handler to transparently accept requests
+// regardless of whether the caller submits an envelope.
+func (b binaryProtocol) DecodeRequest(r io.ReaderAt) (wire.Value, EnvelopeSpecificResponder, error) {
+	// Strip request envelopes if present.
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
+	// without a version (non-strict)
```

Comment:
can you provide rationale here. e.g., why can't it be a valid struct (since field type starts a struct, and 0 is not a valid field type).